lib/HLSL/DxilScalarizeVectorLoadStores.cpp

```diff
@@ -0,0 +1,239 @@
+///////////////////////////////////////////////////////////////////////////////
+//                                                                           //
+// DxilScalarizeVectorLoadStores.cpp                                         //
+// Copyright (C) Microsoft Corporation. All rights reserved.                 //
+// This file is distributed under the University of Illinois Open Source     //
+// License. See LICENSE.TXT for details.                                     //
+//                                                                           //
+// Lowers native vector load stores to potentially multiple scalar calls.    //
+//                                                                           //
+///////////////////////////////////////////////////////////////////////////////
+
+#include "dxc/DXIL/DxilInstructions.h"
+#include "dxc/DXIL/DxilModule.h"
+#include "dxc/HLSL/DxilGenerationPass.h"
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Pass.h"
+
+using namespace llvm;
+using namespace hlsl;
+
+class DxilScalarizeVectorLoadStores : public ModulePass {
+private:
+  DxilModule *m_DM;
+
+  void scalarizeVectorLoad(hlsl::OP *HlslOP, const DataLayout &DL,
+                           CallInst *CI);
+  void scalarizeVectorStore(hlsl::OP *HlslOP, const DataLayout &DL,
+                            CallInst *CI);
+
+public:
+  static char ID; // Pass identification, replacement for typeid
+  explicit DxilScalarizeVectorLoadStores() : ModulePass(ID) {}
+
+  StringRef getPassName() const override {
+    return "DXIL scalarize vector load/stores";
+  }
+
+  bool runOnModule(Module &M) override {
+    DxilModule &DM = M.GetOrCreateDxilModule();
+    m_DM = &DM;
+
+    // Shader Model 6.9 allows native vectors and doesn't need this pass.
+    if (DM.GetShaderModel()->IsSM69Plus())
+      return false;
+
+    bool Changed = false;
+
+    hlsl::OP *HlslOP = DM.GetOP();
+    for (auto F = M.functions().begin(), E = M.functions().end(); F != E;) {
```

Comment:
You can iterate the overloads for specific ops you want instead with `HlslOP::GetOpFuncList`.  For instance:
https://github.com/microsoft/DirectXShaderCompiler/blob/2f357a9d625eaaa982ce1fac513e5f77a7d81900/lib/HLSL/DxilPreparePasses.cpp#L398

Reply:
Oh, for one thing, erasing a function during iteration of this list will erase the entry from the list.  Since this is a SmallMapVector, the iterator points into the vector, which will be invalidated by any modification.  That doesn't seem like it should result specifically in duplicate entries and double-delete, but it is in UB territory.

If the original list contains duplicate entries, that is a cause for concern.  We shouldn't allow that.